### PR TITLE
Fix #4460: Add Wiki documentation for adding the spotlight experience to any element 

### LINKED ISSUE #4460: Add Wiki documentation for adding the spotlight experience to any element

### PR DESCRIPTION
**Pull request by JishnuGoyal**
## Explanation

Fixes #4460 
This PR adds a new page to the Wiki explaining how Spotlights should be used, how new ones should be added and how should they be tested.

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

**Comment by JishnuGoyal** [comment]
Additionally, here is this link to the original Spotlight Wiki doc: https://docs.google.com/document/d/1houcOoZLbTgdwnL1HnmocolEjP6exfY-R_qsNbYPY7Y/edit#

File: wiki/Spotlight Guide
Comment (BenHenning) [inline wiki/Spotlight Guide:1]: This file needs to end with '.md'.

**Review by BenHenning — COMMENTED** [review]
Thanks @JishnuGoyal! I think the filename is incorrect. Also we need to add this to the _Sidebar.md file so that it can be found on the side.

Besides that, since I've already reviewed the contents of this file in detail, I'm actually going to defer to @seanlip for the PR review since he's also broadly going to be reviewing wiki pages moving forward.

**Comment by BenHenning** [comment]
@JishnuGoyal please assign to @seanlip once the comments above are addressed.

File: wiki/Spotlight Guide
Comment (JishnuGoyal) [reply wiki/Spotlight Guide:1]: done

**Comment by JishnuGoyal** [comment]
Hey @BenHenning, how exactly should I make sure this file is shown in the sidebar?

**Comment by JishnuGoyal** [comment]
PTAL @seanlip , thanks!

File: wiki/Spotlight Guide.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
+Feature Spotlights
+# Introduction
+
+Spotlighting is a visual tool that highlights and brings a user’s focus to an element on the screen. We use them to communicate the purpose of a specific screen element to the user, which they might otherwise miss. Spotlighting involves dimming the other areas on the screen and lighting the element to bring the user’s focus onto it, and explaining what/how that feature should be used.
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: At the start: "Spotlights are a visual tool that ..."

Maybe, below this paragraph, add a new one that's like: "Here is an example of a spotlight in the Home screen:" (or wherever) and add a picture to show what a spotlight looks like in practice.

File: wiki/Spotlight Guide.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
+Feature Spotlights
+# Introduction
+
+Spotlighting is a visual tool that highlights and brings a user’s focus to an element on the screen. We use them to communicate the purpose of a specific screen element to the user, which they might otherwise miss. Spotlighting involves dimming the other areas on the screen and lighting the element to bring the user’s focus onto it, and explaining what/how that feature should be used. 
+
+# Creating a new spotlight
+
+The [SpotlightFragment](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/app/src/main/java/org/oppia/android/app/spotlight/SpotlightFragment.kt#L44) contains APIs to call spotlights for any in-app screen element. To create a spotlight, call one of these functions:
+[requestSpotlight](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/app/src/main/java/org/oppia/android/app/spotlight/SpotlightManager.kt#L26)
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: This isn't rendering properly, you may need to put `- ` at the start of this and the next line to get bullet points to show up in markdown.

File: wiki/Spotlight Guide.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+[requestSpotlightWithDelayedLayout](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/app/src/main/java/org/oppia/android/app/spotlight/SpotlightManager.kt#L13)
+
+Both these functions expect a parameter, [SpotlightTarget](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/app/src/main/java/org/oppia/android/app/spotlight/SpotlightTarget.kt#L14). The spotlight target is a holder for the necessary information to show a spotlight. 
+
+The spotlight target binds these fields together:
+anchor: The view that should be spotlit.
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: Ditto for these fields. Please use the markdown preview functionality in GitHub to check your work.

File: wiki/Spotlight Guide.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+hint: The helpful text that should appear along the spotlight to describe the element being spotlit.
+shape: The preferred shape of the spotlight highlight. Can be one of a circle or rounded rectangle, based on whichever shape best fits the area being highlighted
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: one of --> either

End with a period.

File: wiki/Spotlight Guide.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+feature: The specific app feature that the spotlight is tied to. It’s used to track whether this specific spotlight has been seen.
+
+In order to start spotlighting an element in the UI, three high-level things need to be done:
+The feature’s spotlight needs to be defined.
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: Same thing re bullet point formatting.

File: wiki/Spotlight Guide.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+  // Corresponds to the topic fragment's revision tab.
+  SpotlightViewState topic_revision_tab = 3;
+
+  // Similarly, add storage for your new feature here.
+  SpotlightViewState your_feature_name = <next_available_index>;
+}
+```
+
+The new spotlight feature also needs to be added to the [SpotlightStateController](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/domain/src/main/java/org/oppia/android/domain/spotlight/SpotlightStateController.kt#L3). In the [retrieveSpotlightViewState](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/domain/src/main/java/org/oppia/android/domain/spotlight/SpotlightStateController.kt#L80) and the [recordSpotlightStateAsync](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/domain/src/main/java/org/oppia/android/domain/spotlight/SpotlightStateController.kt#L110) method, add the feature to the switch case as so:
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: method --> methods

File: wiki/Spotlight Guide.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+```
+
+Similarly for the [recordSpotlightStateAsync](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/domain/src/main/java/org/oppia/android/domain/spotlight/SpotlightStateController.kt#L110) method.
+
+## Hooking up the spotlight
+In order to request a spotlight in the UI, use [SpotlightManager](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/app/src/main/java/org/oppia/android/app/spotlight/SpotlightManager.kt#L4):
+
+```kotlin
+checkNotNull(getSpotlightManager()).requestSpotlight(
+	SpotlightTarget(
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: Deindent this by 2 perhaps? This should be in a separate indentation block from the next 4 lines.

File: wiki/Spotlight Guide.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	binding.yourFeatureName,
+	R.string.your_feature_string_string_resource,
+	SpotlightShape.RoundedRectangle,
+	Spotlight.FeatureCase.YOUR_FEATURE_NAME
+)
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: Indent this (by whatever is needed to match the corresponding '(' above)

File: wiki/Spotlight Guide.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+  private fun getSpotlightManager(): SpotlightManager? {
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: Make sure the indentation of this matches checkNotNull() above

File: wiki/Spotlight Guide.md [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    val profileId = ProfileId.newBuilder().setInternalId(internalProfileId).build()
+    spotlightStateController.markSpotlightViewed(profileId, TOPIC_LESSON_TAB)  
+    spotlightStateController.markSpotlightViewed(profileId, TOPIC_REVISION_TAB) 
+    spotlightStateController.markSpotlightViewed(profileId, YOUR_FEATURE_NAME)
+    testCoroutineDispatchers.runCurrent()
+  }
+```
+
+## Disabling the spotlights
+If the spotlights need to be disabled, the platform parameter value should be set.
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: This results in the reader asking the question: "Which platform parameter value?"

Either make the next couple of points bullet points, or make them separate paragraphs and use bold highlighting for the initial phrase (e.g. **In tests:** ..., **In production:** ...).

File: wiki/Spotlight Guide.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+If the spotlights need to be disabled in the tests, use the ``TestPlatformParameterModule`` to turn off the [enableSpotlightUi](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/utility/src/main/java/org/oppia/android/util/platformparameter/PlatformParameterConstants.kt#L218) platform parameter.
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: Should this be a new paragraph? If so, add a newline above this one. Please check and do a self-review using the Markdown preview functionality on GitHub.

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks @JishnuGoyal! Took a full pass, PTAL.

Re including this page in the sidebar, @BenHenning's already given you a pointer -- please add it to the relevant place in `_Sidebar.md`.   Thanks!

File: wiki/Spotlight Guide.md (hunk 2 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done

File: wiki/Spotlight Guide.md (hunk 3 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done

File: wiki/Spotlight Guide.md (hunk 4 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done

File: wiki/Spotlight Guide.md (hunk 5 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done

File: wiki/Spotlight Guide.md [hunk 12]
```diff
@@ -6,21 +6,21 @@ Spotlighting is a visual tool that highlights and brings a user’s focus to an
...
 Both these functions expect a parameter, [SpotlightTarget](https://github.com/oppia/oppia-android/blob/d2c37dc547f3e5d12dfe62fa97b9b16fbf0fed6e/app/src/main/java/org/oppia/android/app/spotlight/SpotlightTarget.kt#L14). The spotlight target is a holder for the necessary information to show a spotlight. 
 
 The spotlight target binds these fields together:
-anchor: The view that should be spotlit.
-hint: The helpful text that should appear along the spotlight to describe the element being spotlit.
-shape: The preferred shape of the spotlight highlight. Can be one of a circle or rounded rectangle, based on whichever shape best fits the area being highlighted
-feature: The specific app feature that the spotlight is tied to. It’s used to track whether this specific spotlight has been seen.
+- anchor: The view that should be spotlit.
+- hint: The helpful text that should appear along the spotlight to describe the element being spotlit.
+- shape: The preferred shape of the spotlight highlight. Can either a circle or a rounded rectangle, based on whichever shape best fits the area being highlighted.
```
Comment (seanlip) [inline wiki/Spotlight Guide.md]: either --> be either

File: wiki/Spotlight Guide.md (hunk 1 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: Done - I'm wondering if there's a way to make this image look smaller?

File: wiki/Spotlight Guide.md (hunk 12 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done

File: wiki/Spotlight Guide.md (hunk 11 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done, PTAL

File: wiki/Spotlight Guide.md (hunk 7 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done, PTAL

File: wiki/Spotlight Guide.md (hunk 8 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done, PTAL

File: wiki/Spotlight Guide.md (hunk 9 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done, PTAL

File: wiki/Spotlight Guide.md (hunk 10 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done

File: wiki/Spotlight Guide.md (hunk 6 above)
Comment (JishnuGoyal) [reply wiki/Spotlight Guide.md]: done

**Comment by JishnuGoyal** [comment]
Please take another pass at this @seanlip , thanks!

**Review by seanlip — CHANGES_REQUESTED** [review]
Thanks @JishnuGoyal, I think the page content looks fine. But I think you missed @BenHenning's comment from before. Could you please:

(a) Add this page to the _Sidebar.md file so that users can get to it
(b) Make sure this page is named in the same way as others in the wiki (no spaces in the name, use hyphens)

Once you've done that, this PR can be merged. Thanks!

**Comment by JishnuGoyal** [comment]
Updated the file name and added the same to the sidebar. 

@seanlip PTAL, thanks!

**Review by seanlip — APPROVED** [review]
Thanks! LGTM.

File: wiki/Spotlight Guide.md (hunk 1 above)
Comment (seanlip) [reply wiki/Spotlight Guide.md]: Probably the easiest is to use a smaller image? Feel free to do that in a separate PR though.

**Comment by JishnuGoyal** [comment]
I have updated this branch. @BenHenning PTAL, thanks!

**Comment by seanlip** [comment]
Merged. Thanks @JishnuGoyal!

**Pull request merged**